library/items/item-group/src/main/java/org/quiltmc/qsl/itemgroup/api/QuiltItemGroupBuilder.java

```diff
@@ -0,0 +1,104 @@
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.itemgroup.api;
+
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import net.minecraft.item.ItemGroup;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.collection.DefaultedList;
+import net.minecraft.util.Identifier;
+
+import org.quiltmc.qsl.itemgroup.impl.ItemGroupExtensions;
+
+public final class QuiltItemGroupBuilder {
+	private Identifier identifier;
+	private Supplier<ItemStack> stackSupplier = () -> ItemStack.EMPTY;
+	private Consumer<List<ItemStack>> stacksForDisplay;
+
+	private QuiltItemGroupBuilder(Identifier identifier) {
+		this.identifier = identifier;
+	}
+
+	/**
+	 * Create a new {@link QuiltItemGroupBuilder}.
+	 *
+	 * @param identifier the {@link Identifier} will become the name of the {@link ItemGroup} and will be used for the translation key
+	 * @return a {@link QuiltItemGroupBuilder}
+	 */
+	public static QuiltItemGroupBuilder create(Identifier identifier) {
+		return new QuiltItemGroupBuilder(identifier);
+	}
+
+	/**
+	 * This is used to add an icon to the {@link ItemGroup}.
+	 *
+	 * @param stackSupplier the supplier should return the {@link ItemStack} that you wish to show on the tab
+	 * @return this
+	 */
+	public QuiltItemGroupBuilder icon(Supplier<ItemStack> stackSupplier) {
+		this.stackSupplier = stackSupplier;
+		return this;
+	}
+
+	/**
+	 * This allows for a custom list of items to be displayed in a tab, this enabled tabs to be created with a custom set of items.
+	 *
+	 * @param stacksForDisplay Add {@link ItemStack}s to this list to show in the {@link ItemGroup}
+	 * @return {@code this}
+	 */
+	public QuiltItemGroupBuilder appendItems(Consumer<List<ItemStack>> stacksForDisplay) {
+		this.stacksForDisplay = stacksForDisplay;
+		return this;
+	}
+
+	/**
+	 * This is a single method that makes creating an {@link ItemGroup} with an icon one call.
+	 *
+	 * @param identifier    the id will become the name of the {@link ItemGroup} and will be used for the translation key
+	 * @param stackSupplier the supplier should return the {@link ItemStack} that you wish to show on the tab
+	 * @return An instance of the built {@link ItemGroup}
+	 */
+	public static ItemGroup build(Identifier identifier, Supplier<ItemStack> stackSupplier) {
+		return new QuiltItemGroupBuilder(identifier).icon(stackSupplier).build();
+	}
+
+	/**
+	 * @return An {@link ItemGroup} built from {@code this}
+	 */
+	public ItemGroup build() {
+		((ItemGroupExtensions) ItemGroup.GROUPS[0]).expandArray();
+		return new ItemGroup(ItemGroup.GROUPS.length - 1, String.format("%s.%s", identifier.getNamespace(), identifier.getPath())) {
+			@Override
+			public ItemStack createIcon() {
+				return stackSupplier.get();
+			}
+
+			@Override
+			public void appendStacks(DefaultedList<ItemStack> stacks) {
+				if (stacksForDisplay != null) {
+					stacksForDisplay.accept(stacks);
```

Comment:
We probably need to figure out a better way to handle appending of extra item stacks. At the moment setting the supplier on the builder removes the original contents of the item group.

Reply:
Yeah that's pretty much what I mean, although if we can find the place to "freeze" the registries then it would be better

Reply:
I have a couple issues with the above idea. Mostly that there would then be a dependency on the item-group API from the item API. As well, I think that supplier `ItemGroup`s are fine and make it clearer that stuff needs to be "lazy". Im not sure about the best way to set the icon. But a subclass of item group that allows for setting the icon directly might be better. (Im going to implement this now)

Reply:
@Haven-King take a look at the javadocs and the expanded test mod. I think it balances your suggestion along with making it simple for users and keeping the API implementation simple

Reply:
@i509VCB I made it so that it no longer returns, allowing for items to be added and item stacks.